### PR TITLE
feat(Letterboxd): add presence

### LINKED ISSUE #5230: Letterboxd | letterboxd.com
### Discussed in https://github.com/PreMiD/Presences/discussions/4471

<div type='discussions-op-text'>

<sup>Originally posted by **fl34b4g** June 16, 2020</sup>
**Prerequisites and essential questions** 
- [x] Is it a popular website?
- [x] Is the website older than 2 months? 
- [ ] Is the site locked to a specific country/region?
- [ ] Is the site a paid service? (e.g. Netflix, Hulu)
- [ ] Does the website feature NSFW content? (e.g. porn, etc...)
- [x] Do you acknowledge that coding presences is completely voluntary and may take time for your service to be added regardless of priority?

**What's the name of the service?** 
www.letterboxd.com | Letterboxd

**What should the Presence display?** 
What page you're browsing/viewing, what movie title you're watching.  If possible also what trailer ur watching. </div>

### PR DESCRIPTION
**Pull request by Skaytacium**
**Describe the changes in this pull request:**
Resolves #5230

<details>
<summary> Proof showing the creation/modification is working as expected </summary>
<br>

image
image
image
image
image
image
image
image
image
image
image
image
image
image
image
image
image

</details>

### CODE REVIEW

File: websites/L/Letterboxd/presence.ts [hunk 1]
```diff
@@ -0,0 +1,450 @@
...
+}
+
+presence.on("UpdateData", async () => {
+	const path = document.location.pathname.slice(1).split("/");
+	path.pop();
+
+	const pd: PresenceData = {
+		largeImageKey: "final",
+		startTimestamp: Date.now(),
+		buttons: [{ label: "View", url: window.location.href }]
```
Comment (EncryptedDev) [inline websites/L/Letterboxd/presence.ts:38]: View what...

File: websites/L/Letterboxd/presence.ts [hunk 2]
```diff
@@ -0,0 +1,450 @@
...
+	fnc: (val: T, ind?: number) => boolean
+) {
+	for (let i = 0; i < itr.length; i++) if (fnc(itr[i], i)) return itr[i];
+}
+
+presence.on("UpdateData", async () => {
+	const path = document.location.pathname.slice(1).split("/");
+	path.pop();
+
+	const pd: PresenceData = {
```
Comment (EncryptedDev) [inline websites/L/Letterboxd/presence.ts:35]:
```suggestion
	const presenceData: PresenceData = {
```

**Review by EncryptedDev — APPROVED** [review]

**Review by EncryptedDev — CHANGES_REQUESTED** [review]
Look at my previous comments

**Comment by Skaytacium** [comment]
Added requested changes

<details>
<summary> Screenshots </summary>
<br>

image
image
image
image
image
image


</details>

**Comment by Skaytacium** [comment]
Are any further changes required?

File: websites/L/Letterboxd/presence.ts [hunk 3]
```diff
@@ -0,0 +1,543 @@
...
+					break;
+				}
+		}
+	} else presenceData.details = "At home";
+
+	if (!(await presence.getSetting("show_buttons")))
+		// eslint-disable-next-line no-undefined
+		presenceData.buttons = undefined;
+
+	if (details !== presenceData.details) {
```
Comment (Slowlife01) [inline websites/L/Letterboxd/presence.ts:539]: Pretty unnecessary

**Review by Slowlife01 — DISMISSED** [review]

File: websites/L/Letterboxd/presence.ts (hunk 3 above)
Comment (Skaytacium) [reply websites/L/Letterboxd/presence.ts:538]: Its there for not messing up the `Date.now()` of "Time elapsed"

File: websites/L/Letterboxd/presence.ts [hunk 4]
```diff
@@ -0,0 +1,541 @@
...
+									presenceData.state = `${title}, ${
+										(
+											header.lastElementChild.firstElementChild
+												.firstElementChild as HTMLAnchorElement
+										).innerText
+									}`;
+									presenceData.largeImageKey = getImageURLByAlt(title);
+									presenceData.smallImageKey = "final";
+									// eslint-disable-next-line no-undefined
+									presenceData.startTimestamp = undefined;
```
Comment (Slowlife01) [inline websites/L/Letterboxd/presence.ts:196]:
```suggestion
									delete presenceData.startTimestamp;
```

File: websites/L/Letterboxd/presence.ts [hunk 5]
```diff
@@ -0,0 +1,541 @@
...
+	return str
+		.replaceAll(String.fromCharCode(160), " ")
+		.replaceAll(String.fromCharCode(8217), "'");
+}
+
+function getImageURLByAlt(alt: string) {
+	const imgs = document.getElementsByTagName("img");
+
+	for (let i = 0; i < imgs.length; i++)
+		if (imgs[i].alt === alt) return imgs[i].src;
```
Comment (Slowlife01) [inline websites/L/Letterboxd/presence.ts:30]:
```js
Array.from(document.getElementsByTagName("img")).find(img => img.alt === alt)?.src;
```

File: websites/L/Letterboxd/presence.ts [hunk 6]
```diff
@@ -0,0 +1,541 @@
...
+}
+
+function filterIterable<T extends Element>(
+	itr: HTMLCollectionOf<T>,
+	fnc: (val: T, ind?: number) => boolean
+) {
+	for (let i = 0; i < itr.length; i++) if (fnc(itr[i], i)) return itr[i];
```
Comment (Slowlife01) [inline websites/L/Letterboxd/presence.ts:37]:
```js
Array.from(itr).find((element, ind) => fnc(element, ind));
```

**Review by EncryptedDev — CHANGES_REQUESTED** [review]
Format the metadata to use tabs with Prettier. `prettier --write .`

File: websites/L/Letterboxd/presence.ts [hunk 7]
```diff
@@ -0,0 +1,539 @@
+const presence = new Presence({
+		clientId: "938732156346314795"
+	}),
+	user = document.cookie
+		.split(";")
+		.find(val => val.startsWith("letterboxd"))
+		.split("=")[1];
```
Comment (Bas950) [inline websites/L/Letterboxd/presence.ts:7]:
```suggestion
	user = document.cookie
		.split(";")
		.find(val => val.startsWith("letterboxd"))
		.split("=")[1],
	browsingTimestamp = Math.floor(Date.now() / 1000);
```

File: websites/L/Letterboxd/presence.ts [hunk 8]
```diff
@@ -0,0 +1,539 @@
...
+	return Array.from(itr).find((element, ind) => fnc(element, ind));
+}
+
+presence.on("UpdateData", async () => {
+	const path = document.location.pathname.slice(1).split("/");
+	path.pop();
+
+	const presenceData: PresenceData = {
+		largeImageKey: "final",
+		startTimestamp: Date.now()
```
Comment (Bas950) [inline websites/L/Letterboxd/presence.ts:45]:
```suggestion
		startTimestamp: browsingTimestamp
```

File: websites/L/Letterboxd/presence.ts [hunk 9]
```diff
@@ -0,0 +1,539 @@
...
+					];
+
+					break;
+				}
+		}
+	} else presenceData.details = "At home";
+
+	if (!(await presence.getSetting("show_buttons"))) delete presenceData.buttons;
+
+	if (details !== presenceData.details) {
```
Comment (Bas950) [inline websites/L/Letterboxd/presence.ts:535]: now this can be removed

**Review by Bas950 — CHANGES_REQUESTED** [review]

**Review by Bas950 — APPROVED** [review]

**Review by EncryptedDev — APPROVED** [review]

**Comment by EncryptedDev** [comment]
Thanks for the PR!

**Comment by elijaholmos** [comment]
awesome stuff, thank you!

**Comment by xxKR15Hxx** [comment]
Hey, thanks for the PR really needed to show my friends want I'm doing on letterboxd and you fixed that,
Now all i need is friends

**Pull request merged**